src/screens/SiteDetailScreen.js

```diff
@@ -9,19 +9,85 @@ const {Marker} = MapView;
 
 import _ from 'lodash';
 
+import {attemptToAddFavorite, attemptToRemoveFavorite} from '../actions';
 
 import {linkColorBlue, navyBlueButton} from '../styles/index';
 
-import {navKeys, facilityIconDetails, featureIconDetails} from '../constants';
+import {navKeys, facilityIconDetails, featureIconDetails, map, tokens} from '../constants';
 import {site_detail_screen, campsite, common} from '../locale.en';
 import {campsiteIcon} from "../styles";
 
-const {header_title} = site_detail_screen;
 const {campsite_form} = campsite;
 const {location} = common;
 
 class SiteDetailScreen extends Component {
 
+    componentDidMount() {
+        const {selectedSite, currentUser, navigation: {setParams}} = this.props;
+
+        if (currentUser.name === tokens.GUEST) {
+            return;
+        }
+
+        const isFavorite = !!_.find(currentUser.favorites, favorite => favorite.id === selectedSite.id);
+
+        setParams({selectedSite, isFavorite, currentUser, toggleSiteFavorite: this.toggleSiteFavorite});
+    }
+
+    componentWillReceiveProps(nextProps) {
+        const {navigation: {setParams, state: {params}}} = this.props;
+
+        if (!_.isUndefined(params)) {
+            const {currentUser, selectedSite} = nextProps;
+            const willBeFavorite = !!_.find(currentUser.favorites, favorite => favorite.id === selectedSite.id);
+
+            if(willBeFavorite !== params.isFavorite){
+                setParams({isFavorite: willBeFavorite});
+            }
+        }
+    }
+
+    toggleSiteFavorite = ({isFavorite, selectedSite, currentUser}) => {
+
+        if (!isFavorite) {
+            this.props.attemptToAddFavorite({selectedSite, currentUser});
+        } else {
+            this.props.attemptToRemoveFavorite({selectedSite, currentUser});
+        }
+
+    };
+
+    static renderRightNavButton = ({selectedSite, isFavorite, currentUser, toggleSiteFavorite}) => {
+        if (!currentUser) {
+            return;
+        }
+
+        const {topRightIconStyle} = styles;
+
+        if (Platform.OS === 'ios') {
+            return (
+                <TouchableOpacity style={topRightIconStyle}
+                                  onPress={() => toggleSiteFavorite({isFavorite, selectedSite, currentUser})}>
+                    <Icon type='ionicon'
+                          name={isFavorite ? 'ios-heart' : 'ios-heart-outline'}
+                          size={30}
+                          color={linkColorBlue}
+                    />
+                </TouchableOpacity>
+            );
+        } else if (Platform.OS === 'android') {
+            // android-specific code for navigation here
+        }
+    };
+
+    static navigationOptions = (props) => {
+        const {navigation: {state: {params = {}}}} = props;
+
+        return {
+            headerRight: SiteDetailScreen.renderRightNavButton(params)
+        }
+    };
+
```

Comment:
The code above looks like a lot, but I basically used the same logic we have in other screens to make the heart for favorites dynamic based on the state.